File: leetcode/MaxProfit.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let res = 0;
  let i = 0;
  let peak = prices[0];
  let valley = prices[0];
  while(i < prices.length - 1) {
    while(i < prices.length - 1 && prices[i] >= prices[i+1]) i++
    valley = prices[i]
    while(i < prices.length - 1 && prices[i] <= prices[i+1]) i++
    peak = prices[i]
    res += peak - valley;
  }
  return res
};
/**
 * Best practice
 * var maxProfit = function(prices) {
 *  let res = 0;
 *  for(let i = 1;i<prices.length;i++) {
 *    if(prices[i] > prices[i-1]) {
 *      res += prices[i] - prices[i-1] 
 *    }
 *  }
 *  return res
 * }
 */

File: leetcode/maxProfit.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        return max_profit(prices)


def max_profit(prices):
    min_price = prices[0]
    max_profit = 0
    n = len(prices)
    for i in range(1, n):
        if prices[i] < min_price:
            min_price = prices[i]
        elif prices[i] - min_price > max_profit:
            max_profit = prices[i] - min_price
    return max_profit
